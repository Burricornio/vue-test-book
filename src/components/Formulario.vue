<template>
  <div>
    <form @submit.prevent="onSubmit(inputValue)" action="">
      <input type="text" v-model="inputValue">
      <span class="reversed">{{ reversedInput }}</span>
      <button type="submit">Enviar</button>
    </form>
  </div>
</template>

<script>
import axios from 'axios'

export default {
  props: ['reversed'],
  // forma Cesar con inyecciÃ³n de dependencias, inversiÃ³n de control
  // inject: ['http'],
  data: () => ({
    inputValue: '',
    results: []
  }),
  watch: {
    inputValue (newVal, oldVal) {
      if (newVal.trim().length && newVal !== oldVal) {
        console.log(newVal)
      }
    }
  },
  computed: {
    reversedInput () {
      return this.reversed ? this.inputValue.split('').reverse().join('') : this.inputValue
    }
  },
  methods: {
    onSubmit(value) {
      // Forma Cesar
      // const promesaGet = this.http.get('https://jsonplaceholder.typicode.com/posts?q=' + value)
      const promesaGet = axios.get('https://jsonplaceholder.typicode.com/posts?q=' + value)

      promesaGet.then(results => {
        this.results = results.data
      })

      return promesaGet
    }
  }
}
</script>
