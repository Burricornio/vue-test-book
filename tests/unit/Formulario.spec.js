import { shallowMount } from '@vue/test-utils'
import Formulario from '@/components/Formulario.vue'
import { doesNotReject } from 'assert';

describe('Formulario.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Formulario)
  })

  it('El HTML no ha sufrido modificaciones', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Propiedades', () => {

    it('devuelve la cadena que le pasemos si la propiedad reversed no es true', () => {
      wrapper.setData({ inputValue: 'Probando componente' })

      expect(wrapper.vm.reversedInput).toBe('Probando componente')
    })

    it('devuelve la cadena invertida si la propiedad reversed es true', () => {
      wrapper.setData({ inputValue: 'Hola' })
      wrapper.setProps({ reversed: true })

      expect(wrapper.vm.reversedInput).toBe('aloH')
    })
  })

  describe('Watchers - inputValue', () => {
    let spy

    beforeAll(() => {
      spy = jest.spyOn(console, 'log')
    })

    afterEach(() => {
      // Limpiamos el spy
      spy.mockClear()
    })

    it('no se llama si se pasa valor vacío (sin espacios delante)', () => {
      wrapper.setData({ inputValue: '   ' })
      expect(spy).not.toBeCalled()
    })

    it('no se llama si los valores son iguales', () => {
      wrapper = shallowMount(Formulario, {
        data: () => ({ inputValue: 'Lemmy super perrete' })
      })
      wrapper.setData({ inputValue: 'Lemmy super perrete' })

      expect(spy).not.toBeCalled()
    })

    it('se llama con un nuevo valor en el resto de casos', () => {
      wrapper.setData({ inputValue: 'Lemmy el perrete' })
      expect(spy).toBeCalled()
    })
  })
})
