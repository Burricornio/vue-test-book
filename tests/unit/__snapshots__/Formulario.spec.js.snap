// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formulario.vue El HTML no ha sufrido modificaciones 1`] = `
<div>
  <form
    action=""
  >
    <input
      type="text"
    />
     
    <span
      class="reversed"
    >
      
    </span>
  </form>
</div>
`;
